File: static/intake.js
(()=>{'use strict';const form=document.getElementById('intake-wizard');if(!form)return;
const initial=JSON.parse(form.dataset.budget),pages=[...form.querySelectorAll('[data-intake-page]')],tabs=[...form.querySelectorAll('[data-step]')];let step=0;
const repeaters=[];
form.querySelectorAll('[data-group]').forEach(root=>{const group=root.dataset.group;let rows=Array.isArray(initial[group])?initial[group].map(r=>({...r})):[],index=0;const fields=[...root.querySelectorAll('[data-entry]')],hidden=root.querySelector('input[type=hidden]');
function sync(){if(rows[index])fields.forEach(f=>rows[index][f.dataset.entry]=f.value);hidden.value=JSON.stringify(rows);}
function draw(){root.querySelector('[data-editor]').hidden=!rows.length;root.querySelector('[data-empty]').hidden=!!rows.length;fields.forEach(f=>{f.value=rows[index]?.[f.dataset.entry]??'';f.required=!!rows.length&&(['kind','provider','amount'].includes(f.dataset.entry)||(group==='children'&&['age','married'].includes(f.dataset.entry)));f.disabled=!rows.length;});root.querySelector('[data-count]').textContent=rows.length?`${index+1} / ${rows.length}`:'';root.querySelector('[data-prev]').disabled=index===0;root.querySelector('[data-next]').disabled=index>=rows.length-1;root.querySelector('[data-add]').disabled=rows.length>=100;hidden.value=JSON.stringify(rows);}
function valid(){return fields.every(f=>f.reportValidity());}
root.querySelector('[data-add]').onclick=()=>{sync();if(rows.length&&!valid())return;rows.push({});index=rows.length-1;draw();fields[0].focus();};
root.querySelector('[data-prev]').onclick=()=>{sync();if(valid()){index--;draw();}};root.querySelector('[data-next]').onclick=()=>{sync();if(valid()){index++;draw();}};
root.querySelector('[data-remove]').onclick=()=>{rows.splice(index,1);index=Math.max(0,Math.min(index,rows.length-1));draw();};fields.forEach(f=>f.addEventListener('input',sync));repeaters.push({sync,count:()=>rows.length});draw();});
const stamps=form.elements.foodstamps,stampBox=document.getElementById('foodstamps-amount');function conditional(){const yes=stamps.value==='yes';stampBox.hidden=!yes;form.elements.foodstamps_amount.required=yes;form.elements.foodstamps_amount.disabled=!yes;}stamps.addEventListener('change',conditional);conditional();
const other=form.elements.other_assistance,otherBox=document.getElementById('other-assistance-entries');function otherConditional(){const yes=other.value==='yes';otherBox.hidden=!yes;otherBox.querySelectorAll('[data-entry]').forEach(f=>f.disabled=!yes||f.closest('[data-editor]').hidden);}other.addEventListener('change',otherConditional);otherConditional();
function validPage(i){return [...pages[i].querySelectorAll('input,select,textarea')].every(f=>f.disabled||f.reportValidity());}
function review(){repeaters.forEach(r=>r.sync());const root=document.getElementById('intake-review');root.replaceChildren();const entries=[['children',String(repeaters[0].count())],['rent',form.elements.rent.value],['food',form.elements.food.value],['income',[form.elements.his_income.value,form.elements.her_income.value,form.elements.other_income.value].filter(Boolean).join(' + ')],['assistance',String(repeaters[1].count())],['accounts',String(repeaters[2].count())]];entries.forEach(([key,value])=>{const line=document.createElement('div'),label=document.createElement('span'),v=document.createElement('strong');label.textContent=root.dataset[key];v.textContent=value||root.dataset.blank;line.append(label,v);root.append(line);});const missing=[];if(!form.elements.name_en.value.trim())missing.push('English name');if(!form.elements.name_yi.value.trim())missing.push('Yiddish name');if(!form.elements.address.value.trim()||!form.elements.city.value.trim()||!form.elements.state.value.trim()||!form.elements.zip_code.value.trim())missing.push('Complete address');if(!repeaters[0].count())missing.push('Children');if(!form.elements.rent.value)missing.push('Rent or mortgage');if(!form.elements.food.value)missing.push('Food costs');const list=document.getElementById('readiness-list');list.replaceChildren();(missing.length?missing:['Ready to submit for review']).forEach(item=>{const li=document.createElement('li');li.textContent=item;list.append(li);});}
function show(i){step=i;pages.forEach((p,n)=>p.hidden=n!==i);tabs.forEach((t,n)=>{t.classList.toggle('selected',n===i);t.setAttribute('aria-current',n===i?'step':'false');});document.getElementById('intake-back').disabled=i===0;document.getElementById('intake-next').hidden=i===pages.length-1;document.getElementById('intake-save').hidden=i!==pages.length-1;document.getElementById('intake-progress').textContent=`${i+1} / ${pages.length}`;if(i===pages.length-1)review();}
function move(i){if(i>step&&!validPage(step))return;show(i);}
tabs.forEach(t=>t.onclick=()=>move(Number(t.dataset.step)));document.getElementById('intake-next').onclick=()=>move(step+1);document.getElementById('intake-back').onclick=()=>move(step-1);
form.noValidate=true;form.addEventListener('submit',e=>{repeaters.forEach(r=>r.sync());for(let i=0;i<pages.length;i++){const invalid=[...pages[i].querySelectorAll('input,select,textarea')].find(f=>!f.disabled&&!f.checkValidity());if(invalid){e.preventDefault();show(i);invalid.reportValidity();return;}}});
form.addEventListener('keydown',e=>{if(e.key==='Enter'&&e.target.tagName==='INPUT'&&step<pages.length-1){e.preventDefault();move(step+1);}});show(0);
const totals=document.getElementById('intake-totals');
if(totals){
 const status=document.getElementById('profile-save-status');
 const money=new Intl.NumberFormat(document.documentElement.lang||'en',{style:'currency',currency:'USD'});
 const cents=v=>v!==''&&v!=null&&Number.isFinite(Number(v))&&Number(v)>=0?Math.round(Number(v)*100):null;
 const sum=values=>{const known=values.filter(v=>v!==null);return known.length?known.reduce((a,b)=>a+b,0):null;};
 function updateTotals(dirty=false){
  repeaters.forEach(r=>r.sync());
  const val=key=>form.elements[key].value;
  const accounts=JSON.parse(form.elements.accounts_json.value||'[]');
  const bills=sum(accounts.map(r=>cents(r.monthly_bill)));
  document.getElementById('provider-bills-total').textContent=bills===null?totals.dataset.blank:money.format(bills/100);
  const costs=sum([...['rent','food'].map(k=>cents(val(k))),...accounts.filter(r=>r.budget_treatment==='additional').map(r=>cents(r.monthly_bill))]);
  const income=sum(['his_income','her_income','other_income'].map(k=>cents(val(k))));
  const assistance=JSON.parse(form.elements.assistance_json.value||'[]');
  const help=sum([val('foodstamps')==='yes'?cents(val('foodstamps_amount')):val('foodstamps')==='no'?0:null,...(val('other_assistance')!=='no'?assistance.map(r=>cents(r.amount)):val('other_assistance')==='no'?[0]:[])]);
  const foodCosts=(cents(val('food'))??0)+accounts.filter(r=>r.kind==='grocery'&&r.budget_treatment==='additional').reduce((n,r)=>n+(cents(r.monthly_bill)??0),0);
  const stamps=val('foodstamps')==='yes'?(cents(val('foodstamps_amount'))??0):0;
  const usableHelp=(help??0)-stamps+Math.min(stamps,foodCosts);
  const values={children:repeaters[0].count()||totals.dataset.blank,costs,income,help,gap:costs===null?null:Math.max(0,costs-(income??0)-usableHelp),accounts:JSON.parse(form.elements.accounts_json.value||'[]').length};
  totals.querySelectorAll('[data-total]').forEach(el=>{const key=el.dataset.total,v=values[key];el.textContent=['children','accounts'].includes(key)?v:v===null?totals.dataset.blank:money.format(v/100);});
  if(dirty)totals.dataset.dirty='yes';
  status.textContent=totals.dataset.dirty==='yes'?totals.dataset.unsaved:totals.dataset.existing==='yes'?totals.dataset.saved:totals.dataset.new;
 }
 form.addEventListener('input',()=>updateTotals(true));
 form.addEventListener('change',()=>updateTotals(true));
 form.addEventListener('click',e=>{if(e.target.closest('[data-add],[data-remove]'))updateTotals(true);});
 updateTotals();
}
})();

File: intake.py
"""Validated household budget data. Monetary values are integer cents."""
import json
from decimal import Decimal, InvalidOperation

MONEY_FIELDS = ('rent', 'food', 'his_income', 'her_income', 'other_income', 'foodstamps_amount')

def validate_intake(form):
    def text(key, limit=300):
        value = form.get(key, '').strip()
        if len(value) > limit:
            raise ValueError('Intake text is too long.')
        return value
    def money(value):
        if value in ('', None):
            return None
        try:
            number = Decimal(str(value))
            if not number.is_finite() or number < 0 or number > 1000000 or number.as_tuple().exponent < -2:
                raise ValueError()
            return int(number * 100)
        except (InvalidOperation, ValueError):
            raise ValueError('Enter a valid amount with up to two decimal places, no greater than $1,000,000.')
    data = {k: money(form.get(k, '')) for k in MONEY_FIELDS}
    count = text('children_count')
    if count and (not count.isdecimal() or not 0 <= int(count) <= 50):
        raise ValueError('Number of children must be between 0 and 50.')
    data['children_count'] = int(count) if count else None
    for key in ('his_employment', 'her_employment', 'his_employer', 'her_employer'):
        data[key] = text(key)
    data['foodstamps'] = text('foodstamps')
    if data['foodstamps'] not in ('', 'yes', 'no'):
        raise ValueError('Choose Yes or No for food stamps.')
    if data['foodstamps'] != 'yes':
        data['foodstamps_amount'] = None
    data['other_assistance'] = text('other_assistance')
    if data['other_assistance'] not in ('', 'yes', 'no'):
        raise ValueError('Choose Yes or No for other assistance.')
    for group in ('assistance', 'accounts'):
        try:
            entries = json.loads(form.get(group + '_json', '[]'))
        except (ValueError, TypeError):
            raise ValueError('Invalid account or assistance entry.')
        if not isinstance(entries, list) or len(entries) > 100:
            raise ValueError('Invalid account or assistance entry.')
        result = []
        for entry in entries:
            if not isinstance(entry, dict):
                raise ValueError('Invalid account or assistance entry.')
            allowed = ('provider', 'amount') if group == 'assistance' else ('kind', 'provider', 'account', 'phone', 'child', 'monthly_bill', 'budget_treatment')
            cleaned = {}
            for key in allowed:
                value = entry.get(key, '')
                if value is None: value = ''
                if not isinstance(value, str) or len(value) > 300:
                    raise ValueError('Invalid account or assistance entry.')
                cleaned[key] = value.strip()
            if not any(cleaned.values()): continue
            if group == 'assistance':
                cleaned['amount'] = money(cleaned['amount'])
            elif cleaned['kind'] not in ('', 'utility', 'grocery', 'mosdos', 'other'):
                raise ValueError('Choose an account type.')
            if group == 'accounts':
                cleaned['monthly_bill'] = money(cleaned['monthly_bill'])
                if cleaned['budget_treatment'] not in ('', 'additional', 'food', 'rent'):
                    raise ValueError('Choose how to count this bill')
            result.append(cleaned)
        if group == 'assistance':
            if data['other_assistance'] == 'no':
                result = []
        data[group] = result
    return data


def intake_for_form(data):
    data = dict(data or {})
    for key in MONEY_FIELDS:
        data[key] = '' if data.get(key) is None else f'{data[key] / 100:.2f}'
    data['accounts'] = [dict(row, monthly_bill='' if row.get('monthly_bill') is None else f"{row['monthly_bill'] / 100:.2f}") for row in data.get('accounts', [])]
    data['assistance'] = [dict(row, amount='' if row.get('amount') is None else f"{row['amount'] / 100:.2f}") for row in data.get('assistance', [])]
    return data
